File: app/src/main/java/dev/ragnarok/fenrir/model/WallEditorAttrs.java
package dev.ragnarok.fenrir.model;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

public final class WallEditorAttrs implements Parcelable {

    public static final Creator<WallEditorAttrs> CREATOR = new Creator<WallEditorAttrs>() {
        @Override
        public WallEditorAttrs createFromParcel(Parcel in) {
            return new WallEditorAttrs(in);
        }

        @Override
        public WallEditorAttrs[] newArray(int size) {
            return new WallEditorAttrs[size];
        }
    };
    private final ParcelableOwnerWrapper owner;
    private final ParcelableOwnerWrapper editor;

    public WallEditorAttrs(@NonNull Owner owner, @NonNull Owner editor) {
        this.owner = new ParcelableOwnerWrapper(owner);
        this.editor = new ParcelableOwnerWrapper(editor);
    }

    private WallEditorAttrs(Parcel in) {
        owner = in.readParcelable(ParcelableOwnerWrapper.class.getClassLoader());
        editor = in.readParcelable(ParcelableOwnerWrapper.class.getClassLoader());
    }

    @NonNull
    public Owner getOwner() {
        return owner.get();
    }

    @NonNull
    public Owner getEditor() {
        return editor.get();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(owner, flags);
        dest.writeParcelable(editor, flags);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/WallEditorAttrs.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.readTypedObjectCompat
import dev.ragnarok.fenrir.writeTypedObjectCompat

class WallEditorAttrs : Parcelable {
    private val owner: ParcelableOwnerWrapper
    private val editor: ParcelableOwnerWrapper

    constructor(owner: Owner, editor: Owner) {
        this.owner = ParcelableOwnerWrapper(owner)
        this.editor = ParcelableOwnerWrapper(editor)
    }

    internal constructor(`in`: Parcel) {
        owner = `in`.readTypedObjectCompat(ParcelableOwnerWrapper.CREATOR)!!
        editor = `in`.readTypedObjectCompat(ParcelableOwnerWrapper.CREATOR)!!
    }

    fun getOwner(): Owner {
        return owner.get()!!
    }

    fun getEditor(): Owner {
        return editor.get()!!
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeTypedObjectCompat(owner, flags)
        dest.writeTypedObjectCompat(editor, flags)
    }

    companion object CREATOR : Parcelable.Creator<WallEditorAttrs> {
        override fun createFromParcel(parcel: Parcel): WallEditorAttrs {
            return WallEditorAttrs(parcel)
        }

        override fun newArray(size: Int): Array<WallEditorAttrs?> {
            return arrayOfNulls(size)
        }
    }
}